src/redisearch_rs/rlookup/src/lookup.rs

```diff
@@ -0,0 +1,279 @@
+    ///
+    /// The caller is responsible for the memory previously managed by the `Box`, in particular
+    /// the caller should properly destroy the `RLookupKey` and deallocate the memory by calling
+    /// `Self::from_ptr`.
+    ///
+    /// # Safety
+    ///
+    /// The caller *must* continue to treat the pointer as pinned.
+    #[inline]
+    #[cfg_attr(not(test), expect(unused, reason = "used by later stacked PRs"))]
+    unsafe fn into_ptr(me: Pin<Box<Self>>) -> NonNull<Self> {
+        // This function must be kept in sync with `Self::from_ptr` below.
+
+        // Safety: The caller promised to continue to treat the returned pointer
+        // as pinned and never move out of it.
+        let ptr = Box::into_raw(unsafe { Pin::into_inner_unchecked(me) });
+
+        // Safety: we know the ptr we get from Box::into_raw is never null
+        unsafe { NonNull::new_unchecked(ptr) }
+    }
+
+    /// Constructs a `Pin<Box<RLookupKey>>` from a raw pointer.
+    ///
+    /// The returned `Box` will own the raw pointer, in particular dropping the `Box`
+    /// will deallocate the `RLookupKey`. This function should only be used by [`RLookup::drop`].
+    ///
+    /// # Safety
+    ///
+    /// 1. The caller must ensure the pointer was previously created through [`Self::into_ptr`].
+    /// 2. The caller has to be careful to never call this method twice for the same pointer, otherwise a
+    ///    double-free or other memory corruptions will occur.
+    /// 3. The caller *must* also ensure that `ptr` continues to be treated as pinned.
+    #[inline]
+    #[cfg_attr(not(test), expect(unused, reason = "used by later stacked PRs"))]
+    unsafe fn from_ptr(ptr: NonNull<Self>) -> Pin<Box<Self>> {
+        // This function must be kept in sync with `Self::into_ptr` above.
+
+        // Safety:
+        // 1 -> This function will only ever be called through `RLookup::drop` below.
+        //      We therefore know - because push_key creates pointers through `into_ptr` - that the invariant is upheld.
+        // 2 -> Has to be upheld by the caller
+        let b = unsafe { Box::from_raw(ptr.as_ptr()) };
+        // Safety: 3 -> Caller has to uphold the pin contract
+        unsafe { Pin::new_unchecked(b) }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Make sure that the `into_ptr` and `from_ptr` functions are inverses of each other.
+    #[test]
+    fn into_ptr_from_ptr_roundtrip() {
+        let key = RLookupKey::new(c"test", RLookupKeyFlags::empty());
+        let key = Box::pin(key);
+
+        let ptr = unsafe { RLookupKey::into_ptr(key) };
+        let key = unsafe { RLookupKey::from_ptr(ptr) };
+
+        assert_eq!(*key, RLookupKey::new(c"test", RLookupKeyFlags::empty()));
+    }
+
+    // Assert that creating a RLookupKey with the NameAlloc flag indeed allocates a new string
+    #[test]
+    fn rlookupkey_new_with_namealloc() {
+        let name = c"test";
```

Comment:
Do we need a test with UTF-8 CStr? 😺 to make sure length is byte-length?

Reply:
Good call, added two new test cases for this